### PR TITLE
Combined Removals for PL-v0.40 (Yushao)

### PR DESCRIPTION
**Pull request by JerryChen97**
This PR merges the following removals in a single PR to reduce impact in our CI:
 - Removal of `BasisStatePreparation` 
 - Removes `ancilla` argument in iterative QPE 
 - Removal of `qml.workflow.set_shots`
 - Removals of `QNode.gradient_fn`

Original PRs and respective Shortcut Stories: 
- [PR #6528](https://github.com/PennyLaneAI/pennylane/pull/6528), [sc-77480]
- [PR #6532](https://github.com/PennyLaneAI/pennylane/pull/6532), [sc-77483]
- [PR #6534](https://github.com/PennyLaneAI/pennylane/pull/6534), [sc-77513]
- [PR #6535](https://github.com/PennyLaneAI/pennylane/pull/6535), [sc-77510]

### CODE REVIEW

File: doc/development/deprecations.rst
Comment (JerryChen97) [inline doc/development/deprecations.rst:39]:
```suggestion
* ``Device``, ``QubitDevice``, and ``QutritDevice`` will no longer be imported top level in v0.40.  They instead
  will be available as ``qml.devices.LegacyDevice``, ``qml.devices.QubitDevice``, and ``qml.devices.QutritDevice``
  respectively.

  - Deprecated top level access in v0.39
  - Top level access will be removed in v0.40
```

File: doc/development/deprecations.rst
Comment (JerryChen97) [inline doc/development/deprecations.rst:38]:
```suggestion
  - Will be removed in v0.40
  
```

File: pennylane/workflow/__init__.py
Comment (lillian542) [inline pennylane/workflow/__init__.py:25]:
```suggestion
    ~execute
```

**Review by lillian542 — APPROVED** [review]
One comment, otherwise looks good :)

File: doc/development/deprecations.rst
Comment (mudit2812) [inline doc/development/deprecations.rst:117]: This was updated in the original PR, needs to be updated here too.

File: doc/releases/changelog-dev.md
Comment (mudit2812) [inline doc/releases/changelog-dev.md:63]: Same comment

File: tests/interfaces/test_autograd_qnode.py
Comment (mudit2812) [inline tests/interfaces/test_autograd_qnode.py:546]: `get_gradient_fn` is being deprecated in #6418. I'd suggest syncing with @andrijapau to figure out how to best handle this deprecation/removal.

**Review by mudit2812 — CHANGES_REQUESTED** [review]
It looks like there are mising changes that were requested in the original PRs. I'm happy to approve once those are added to this PR.

File: doc/development/deprecations.rst
Comment (JerryChen97) [reply doc/development/deprecations.rst:117]: Yeah this is so weird. I remember every suggestion has been committed. Probably this is caused by some conflict merging process

Comment (JerryChen97) [inline doc/development/deprecations.rst:117]:
```suggestion
* The ``qml.workflow.set_shots`` helper function has been removed. We no longer interact with the legacy device interface in our code.
  Instead, shots should be specified on the tape, and the device should use these shots.
```

File: doc/releases/changelog-dev.md
Comment (JerryChen97) [inline doc/releases/changelog-dev.md:63]:
```suggestion
* The `qml.workflow.set_shots` helper function has been removed. We no longer interact with the legacy device interface in our code.
  Instead, shots should be specified on the tape, and the device should use these shots.
  [(#6534)](https://github.com/PennyLaneAI/pennylane/pull/6534)
```

File: tests/interfaces/test_autograd_qnode.py
Comment (JerryChen97) [reply tests/interfaces/test_autograd_qnode.py:546]: Oh no this is horrible. I double check for now. Did not notice this before.

File: tests/interfaces/test_autograd_qnode.py
Comment (JerryChen97) [inline tests/interfaces/test_autograd_qnode.py:545]:
```suggestion
```

File: tests/interfaces/test_autograd_qnode.py
Comment (JerryChen97) [reply tests/interfaces/test_autograd_qnode.py:545]: Andrija confirmed the `get_gradient_fn` being in deprecation process for now. Given such situation, there is no way to merge this removal bundle for now since multiple places involve the usage of `get_gradient_fn` in replace of `gradient_fn`. Further replacement with `diff_method` will be added next Monday. For now I just put not merge to block this pr.

Comment (mudit2812) [reply tests/interfaces/test_autograd_qnode.py:545]: Sounds good, thanks Yushao!

Comment (JerryChen97) [reply tests/interfaces/test_autograd_qnode.py]: After discussion with @albi3ro , we realized that the current usage of `get_gradient_fn` actually seems redundant; meanwhile, the testing strategy using spycall is not a very good solution but will be fixed in some improvement plan [sc-72838]. Therefore, in the newest [commit](https://github.com/PennyLaneAI/pennylane/pull/6558/commits/2f42a800c23f85e752ccb8d4ed55e3b6c41bd7c1) I just delete the mentioned asserts as a temporary solution.

File: doc/releases/changelog-dev.md
Comment (mudit2812) [inline doc/releases/changelog-dev.md:84]:
```suggestion
* `QNode.gradient_fn` has been removed. Please use `QNode.diff_method` instead. `QNode.get_gradient_fn` can also be used to
```

**Review by mudit2812 — APPROVED** [review]

**Pull request merged**
